### DISCUSSION TITLE
243

### DISCUSSION THREAD

**Discussion by jojoldu** [jojoldu/blog-comments#546] [Comments]
http://jojoldu.tistory.com/243

**Comment by jojoldu** [comment]
작성자: 일본개발자

굉장히 유익한 글 감사합니다 ㅜㅜ

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

방문해주셔서 고맙습니다^^

**Comment by jojoldu** [comment]
작성자: 이준수

좋은 글이네요.. 잘 읽고 갑니다.

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

방문해주셔서 고맙습니다^^

**Comment by jojoldu** [comment]
작성자: 우동plus

좋은글 감사합니다. 
'프로그래머 열정을 말하다' 이 책도 한번 읽어봐야 겠네요~~

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

넵^^ 개인적으로는 정말 괜찮은 개발자 에세이 책이라 생각합니다.
꼭 읽어보시길 추천드립니다

**Comment by hankkuu** [comment]
내용이 맞는지 궁금한게 있는데 
먼저 말씀드릴 것은 1개의 컬럼만 인덱스를 걸어야 한다면, 해당 컬럼은 카디널리티(Cardinality)가 가장 높은 것을 잡아야 한다는 점입니다.

이 말이 맞는건가요?? 카디널리티가 높으면 선택도가 높아서 중복컬럼이 많은 경우인데 
중복 컬럼이 많다면 Range Scan에서 걸러지는 경우가 아니라 Full Scan이 될 것 같은데 
선택도가 작은 컬럼을 인덱스로 써야 한다고 해야 맞지 않나요??

-- 이말은 좀 수정해야 겠네요 카디널리티가 높다는 것은 중복이라기 보다는 결과 컬럼이 많다는 것이고 결과 컬럼이 많으니 Full Scan이 될 수 있다고가 더 맞겠네요... 
선택도가 작은 컬럼을 인덱스로 쓴다는 것은 일단 중복 검색될 일이 적으니 인덱스 범위 스캔으로 검색될 가능성이 높다.. 이렇게 알고 있어서.. 위의 내용은 확인을 해보고 싶습니다.

**Comment by jojoldu** [comment]
@hankkuu 음 일단 카디널리티가 높다 == 중복도가 낮다 == 선택도가 높다
인데요
여기서 중복컬럼이 많다는 얘기는 어떤 의미인지 잘모르겠습니다.
예를 들어 성별 vs 주민등록번호가 있다면 여기서 **카디널리티가 높은 컬럼은 주민등록번호**입니다.

여기서 성별을 인덱스로 잡게되면, 사실상 인덱스의 역할을 못하게 됩니다.
50%를 전체스캔하게 되서 그렇구요.
주민등록번호로 인덱스를 잡게 되면, 이는 모든 데이터에서 하나밖에 없는값이기 때문에
인덱스에서 즉시 찾을수 있게 됩니다.

선택도가 높을수록 인덱스만으로 탐색이 되기 때문에 훨씬 효율이 좋습니다.
선택도가 낮은 컬럼을 인덱스로 잡으시면 안됩니다.

[참고](http://www.sqler.com/558930)

**Comment by hankkuu** [comment]
저랑... 알고있는게 반대인것 같네요 선택도가 높으면 확률이 높다이기 때문에 당연히 중복이 높아지는 경우이고요 카디널리티는 비율적 측면이 아닌 해당 조건 검색시 발견되는 집합의 개수로 알고 있습니다. 즉 카디널리티는 어떻게 검색하느냐에 따라 비용적으로 달라진다는 것입니다. 음... 뭐가 맞는 것일까요..

**Comment by jojoldu** [comment]
@hankkuu 음..? 그럼 말씀하신대로라면 성별 vs 주민등록번호에서 성별을 단일 인덱스로 잡는게 더 성능이 좋다는 말씀이신가요?

**Comment by hankkuu** [comment]
아니요 성별은 0.5의 선택도이니 (링크에서는 밀도이네요) 결과 집합은 전체 * 0.5 = 카디널리티 수 
이고 주민등록번호는 못해도 0.5 보다는 훨신 작겠죠 즉 선택도가 작다는 건 확률이나 비율적인 의미이지 실제 수량은 아니라는 것 입니다... 적어도 제가 알기로는 다만 통계나 히스토그램을 이용하는 옵티마이저에서는 선택도나 카디널리티를 다 고려한 실행계획을 나가겠지요 
(인덱스가 있더라도 결과집합이 엄청 많으면 인덱스와 상관없이 풀 스캔이므로 어떤 것을 인덱스로 거는지는 일단 이 Case에서는 제외하겠습니다.)

**Comment by jojoldu** [comment]
음 제가 말씀하신것을 잘 이해했는지 모르겠지만,

일단 첫 댓글에 다음과 같이 남겨주셨는데요.

```
카디널리티가 높으면 선택도가 높아서 중복컬럼이 많은 경우인데
중복 컬럼이 많다면 Range Scan에서 걸러지는 경우가 아니라 Full Scan이 될 것 같은데
선택도가 작은 컬럼을 인덱스로 써야 한다고 해야 맞지 않나요??
```

일단 카디널리티가 높은걸 인덱스로 잡아야 하는게 맞다고 생각합니다.

![image](https://user-images.githubusercontent.com/7760496/38811668-15648826-41c5-11e8-980d-f95c42762a82.png)

[dbguide.net](http://www.dbguide.net/knowledge.db?cmd=view&boardUid=142699&boardConfigUid=30&boardStep=1)

그리고 선택도가 높다 == 카디널리티가 높다에는 동의하시는건가요?
혹시나 제가 둘의 관계를 잘못이해하고 있는지 확인중인데요.

![image](https://user-images.githubusercontent.com/7760496/38812152-6e960c02-41c6-11e8-82aa-db2f0519841b.png)

[다른 블로그](http://12bme.tistory.com/138) 에서도 저와 비슷하게 이해하고 계신것 같아서요.

결과적으로 확인하고 싶으신것이

* 선택도가 낮은 (혹은 카디널리티가 낮은) 컬럼을 인덱스로 잡아야 한다

가 맞으신가요?

**Comment by hankkuu** [comment]
http://wiki.gurubee.net/pages/viewpage.action?pageId=4948754 
구루비 사이트를 보면 선택도는 0 ~ 1 사이 입니다. 일단 수랑 비율(%) 단위 자체가 다르네요 
위의 경우는 100개의 값중에 유니크한 경우가 10이니 
선택도는 10/100 = 0.1 이 되겠네요 
이 상황에서 중복이 더 생긴다면 분자가 커지니 선택도(기수성)이 높아지겠네요 선택도는 높아진다고가 맞겠네요 선택도가 높아지면 검색대상이 많이지니 더 빨리진다는게 이상하네요... 

일단 카디널리티는 전체 * 0.1 = 10으로서 유니크한 개수 즉 유니크한 결과 집합 개수와 같게 나오네요

**Comment by hankkuu** [comment]
http://www.mysqlkorea.com/gnuboard4/bbs/board.php?bo_table=community_07&wr_id=9
아직 정확하지는 않지만.. 약간 착오가 있는 것 같습니다. 
선택도는 일단 0~1구간인데 이 때 1인 선택도는 모든 데이터가 중복이 없다고 볼 수 있습니다. 
즉 이때 전체 개수 *  컬럼의 선택도 (1) = 전체 Row (해당 컬럼의 카디널리티) 

이 공식에 의하면 카디널리티가 커야 선택도가 크다는 말이네요 
즉 선택도가 큰 컬럼은 인덱스로 잡아야 하는 것이 맞고 그 상황에서 카디널리티는 다른 컬럼일 때보다 크게 됩니다. 

이러면 알고계신 거랑 같아 지는 건가요??
다만 http://mapu.tistory.com/38
여기는 또 반대입니다. ... 허허..

**Comment by hankkuu** [comment]
공부를 진행해보니 위의 내용이 맞는 쪽으로 결론이 났습니다 
도움주셔서 감사합니다

**Comment by jojoldu** [comment]
@hankkuu 아! 저도 이번기회로 더 자세히 공부하게됐습니다ㅎ
건전한 토론 감사합니다!

**Comment by baeharam** [comment]
창천향로님, 항상 자극받고 있습니다. 위에서 카디널리티에 대한 정의를 "중복된 수치"라고 하셨는데 그러면 성별은 남/여 밖에 없으니까 중복된 수치가 높아서 카디널리티가 높은 것이 아닌가요? 조금 더 명확한 설명 부탁드려도 될까요?

**Comment by ryudung** [comment]
시간들여 써주신 좋은 글 감사합니다.
오늘도 배워갑니다!

**Comment by jojoldu** [comment]
@baeharam 앗 이제 봤습니다 ㅠ 답변 늦어서 죄송합니다 ㅠ
카디널리티를 이해하실때 중복도/선택도 이런 단어는 오히려 오해의 소지가 크구요
(위에서 토론에서 hankkuu님께서 중복도와 선택도 단어를 선택하셔서 같이 사용하긴 했습니다만.)
보통 제가 이해할때는 **해당 컬럼의 유니크값의 수**로 이해하고 사용합니다.
카디널리티 중복도 선택도 란 단어에 함몰되면 오히려 이론에 빠져서 실전에서 못쓸일이 많아서요

유니크 키값의 수가 많은 것 -> 적은 순으로 인덱스를 잡아야 한다 
정도로만 이해하시면 충분하실것 같습니다

[RealMysql 저자분 블로그](http://intomysql.blogspot.com/2010/12/myisam-innodb.html) 도 함께 보시면 도움되실것 같습니다!

**Comment by jojoldu** [comment]
@ryudung ㅋㅋㅋ 감사합니다 :) 영재씨도 갔으니 한번 뵈야죠!?

**Comment by baeharam** [comment]
@jojoldu 테스트 할 때 DISTINCT 쿼리 쓰는걸 캐치 못했네요 ㅋㅋㅋ 자세한 설명 감사합니다!! 그리고 좋은 책의 문구도 감사합니다! 저런 스타일의 책을 너무 좋아하는데 꼭 읽어보겠습니다!ㅎㅎ

**Comment by not-for-me** [comment]
많이 배우고 갑니다!
귀한 컨텐츠 작성 감사합니다!

**Comment by devveb** [comment]
시간이 지나도 유용한 훌륭한 글이네요. 감사합니다.

**Comment by Y-k-Y** [comment]
정말 유익한 글이네요 좋은 글 언제나 감사합니다.👍

**Comment by hsoh-neibus** [comment]
인덱스를 디비가 어떻게 다루는지 이해하는데 큰 도움이 되었습니다!

**Comment by joshua1988** [comment]
동욱님 넘 멋있어요 잘 읽었습니다 :)

**Comment by Jongmun-Park** [comment]
잘 봤습니다 :)

**Comment by kor-Chipmunk** [comment]
좋은 글 감사합니다.

**Comment by homekeeper89** [comment]
진짜 좋은글이네요

**Comment by JongHoo** [comment]
잘읽었습니다 감사합니다~~

**Comment by writer0713** [comment]
좋은 정보 감사합니다. 읽다가 중간에 궁금한점이 생겨서 댓글 남깁니다. 
“카디널리티가 높은순 -> 낮은순” 으로 하는것이 성능상 더 좋다. 라는 부분이 위에서 설명할때 사용한 인덱스와 컬럼들에 해당되는 내용인가요? 아니면, 다른 상황에서도 해당되는 내용인가요?

**Comment by YeomJaeSeon** [comment]
'인덱스 적용했는데 왜 성능 향상이 안되지?' 등의 고민이 있었는데 글보고 해결했습니다. 좋은글 감사합니다.

**Comment by xeounxzxu** [comment]
시간이 지나도 좋은 설명 같습니다. 
인덱스를 사용을 해서 튜닝을 하면서 궁금했던 점 과 카디널리티 순서 그리고 인덱스를 왜 걸어야 하며, Real Mysql 을 보고 난 뒤 동욱님 글 보니 더 많이 이해가 되는 것 같습니다. 
항상 감사합니다.

**Comment by imdudu1** [comment]
보물상자.. 잘 읽었습니당!!

**Comment by jys9962** [comment]
오래된 글이지만 제가 아는 내용과 달라서 질문겸 댓글을 달아봅니다...
"여러 컬럼으로 인덱스를 잡는다면 카디널리티가 높은순에서 낮은순으로 (group_no, from_date, is_bonus) 구성하는게 더 성능이 뛰어납니다" 부분에서

동일 테이블에 인덱스의 컬럼 순서만 다른 경우
두 인덱스의 크기가 같을 것이고 인덱스 조회시 B트리의 높이(LogN)가 같을 것이므로 성능은 동일해야 할 것 같습니다
위 실험에서 결과가 다른 이유는 실행계획을 봐야겠지만 다른 이유일 것 같습니다 (group_no가 in절이라서거나...)

**Reply by galmegiz** [reply]
저도 최근에 읽은 친절한 오라클 SQL튜닝 책에서도 그렇고 말씀하신 B-Tree의 개념을 생각했을 때 차이가 거의 없어야 할 것 같은데 쿼리 실행 결과가 저렇게 뚜렷하게 차이가 나는 이유를 모르겠네요
심지어 상황에 따라서는 low cardinality 칼럼을 선두컬럼으로 두는 게 나을 때도 있다고 하니 저도 한 번 테스트를 해봐야겠습니다.
